util/include/MultiSizeBufferPool.hpp

```diff
@@ -0,0 +1,483 @@
+    void onChunkDeleted(uint64_t chunkSize);
+    void onInit(uint32_t numMaxBuffers, uint32_t chunkSize);
+    void setLimits(uint64_t maxAllocatedBytes);
+
+    struct CurrentStateStatistics {
+      // Pool level:
+      // numAllocatedBytes_ + numNonAllocatedBytes_ = Config::maxAllocatedBytes + numAllocatedMetadataBytes_ (if non
+      // zero) in pool level.
+      // Subpool level:
+      // numAllocatedBytes_ + numNonAllocatedBytes_ = SubpoolConfig::numMaxBuffers * (SubpoolConfig::bufferSize +
+      // chunkMetadataSize())
+      std::atomic_uint64_t numAllocatedBytes_{};
+      std::atomic_uint64_t numNonAllocatedBytes_{};
+
+      // numAllocatedBytes_ = numUnusedBytes_ + numUsedBytes_
+      std::atomic_uint64_t numUsedBytes_{};
+      std::atomic_uint64_t numUnusedBytes_{};
+
+      // numAllocatedChunks_ + numNonAllocatedChunks_ = SubpoolConfig::numMaxBuffers
+      std::atomic_uint64_t numAllocatedChunks_{};
+      std::atomic_uint64_t numNonAllocatedChunks_{};
+
+      // numAllocatedChunks_ = numUsedChunks_ + numUnusedChunks_
+      std::atomic_uint64_t numUsedChunks_{};
+      std::atomic_uint64_t numUnusedChunks_{};
+    } current_;
+
+    // Overall statistics are accumulating
+    struct OverallStatistics {
+      std::atomic_uint64_t numAllocatedBytes_{};
+      std::atomic_uint64_t numDeletedBytes_{};
+      std::atomic_uint64_t numUsedBytes_{};
+
+      std::atomic_uint64_t numAllocatedChunks_{};
+      std::atomic_uint64_t numDeletedChunks_{};
+      std::atomic_uint64_t numUsedChunks_{};
+
+      // Amount of times where a chunk did not fit into the needed buffer size and returned to pool without use
+      std::atomic_uint64_t numBufferUsageFailed_{};
+      // Amount of times where a chunk fit into the needed buffer size and returned to pool without use
+      std::atomic_uint64_t numBufferUsageSuccess_{};
+
+      std::atomic_uint64_t maxNumAllocatedBytes_{};
+      std::atomic_uint64_t maxNumAllocatedChunks_{};
+    } overall_;
+
+    // relevant only if reportOnChangesOnly_ is true. Used to mark if the content need to be reported, only if there was
+    // a change
+    std::atomic_bool reportFlag_{false};
+    // Report only if there were changes. This is in order to prevent endless reports into log when system is idle.
+    const bool enableStatusReportChangesOnly_;
+  };
+
+  friend std::ostream& operator<<(std::ostream& os, const MultiSizeBufferPool::Statistics& s);
+
+  struct ChunkHeader {
+    ChunkHeader(uint32_t bufferSize) : magicField_{kMagicField}, bufferSize_(bufferSize) {}
+    static constexpr int kMagicField = 0xFCFCFCFA;
+    const int magicField_;
+    // buffer size to be used by client, does not include chunkHeaderSize() since this part is fixed.
+    const uint32_t bufferSize_;
+  };
+
+  struct ChunkTrailer {
+    ChunkTrailer() : magicField_{kMagicField} {}
+    static constexpr int kMagicField = 0xFAFAFAFC;
+    const int magicField_;
+  };
+
+  class SubPool {
```

Comment:
Why to define all other classes and struct in MultiSizeBufferPool?

Reply:
Since they should be protected inside the MultiSizeBufferPool. Doing differently is just not right in my opinion.
You can create from the outside ONLY MultiSizeBufferPool, I don't want you to create anything else. Furthermore, Subpool or any other sub-class, as far as I am concerned is not an individual entity, it is part of the pool, and I want to make sure it's fully understood. Having the classes seperated is coming from C style language code writing style.
All classes compose the pool, inside the pool's scope, and protected to prevent anyone to create them directly. 
Caller can create and access a single type (except the ones which are part of API, e.g configuration classes) .
There are other ways to do that (pimpl idiom and more), but this is the most direct and simple one.
Other reasons: 1) shorter names 2) prevent name polution.